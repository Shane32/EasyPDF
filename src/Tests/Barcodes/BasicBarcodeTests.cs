using System.Drawing;

namespace Tests.Barcodes;

public class BasicBarcodeTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public BasicBarcodeTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void QRAlignment()
    {
        using var generator = new QRCoder.QRCodeGenerator();
        using var qr = generator.CreateQrCode("https://github.com/Shane32/EasyPDF", QRCoder.QRCodeGenerator.ECCLevel.L);

        PrintPage(true);
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        PrintPage(false);
        
        void PrintPage(bool quietZone)
        {
            _writer.PictureAlignment = PictureAlignment.LeftTop;
            _writer.MoveTo(0, 0).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(0, 0));
            Box(_writer.QRCodeSize(qr, quietZone: quietZone));
            _writer.Position.ShouldBe(new PointF(0, 0));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.CenterTop;
            _writer.MoveTo(6.5f / 2, 0).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(6.5f / 2, 0));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.RightTop;
            _writer.MoveTo(6.5f, 0).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(6.5f, 0));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.LeftCenter;
            _writer.MoveTo(0, 8f / 2).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(0, 8f / 2));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.CenterCenter;
            _writer.MoveTo(6.5f / 2, 8f / 2).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(6.5f / 2, 8f / 2));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.RightCenter;
            _writer.MoveTo(6.5f, 8f / 2).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(6.5f, 8f / 2));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.LeftBottom;
            _writer.MoveTo(0, 8f).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(0, 8f));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.CenterBottom;
            _writer.MoveTo(6.5f / 2, 8f).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(6.5f / 2, 8f));
            CrossHair(2f);

            _writer.PictureAlignment = PictureAlignment.RightBottom;
            _writer.MoveTo(6.5f, 8f).QRCode(qr, quietZone: quietZone);
            _writer.Position.ShouldBe(new PointF(6.5f, 8f));
            CrossHair(2f);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void BarcodeAlignment()
    {
        var str = "https://github.com";

        _writer.PictureAlignment = PictureAlignment.LeftTop;
        _writer.MoveTo(0, 0).Barcode(str);
        _writer.Position.ShouldBe(new PointF(0, 0));
        Box(_writer.BarcodeSize(str), 0.5f);
        _writer.Position.ShouldBe(new PointF(0, 0));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.CenterTop;
        _writer.MoveTo(6.5f / 2, 1).Barcode(str);
        _writer.Position.ShouldBe(new PointF(6.5f / 2, 1));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.RightTop;
        _writer.MoveTo(6.5f, 2).Barcode(str);
        _writer.Position.ShouldBe(new PointF(6.5f, 2));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.LeftCenter;
        _writer.MoveTo(0, 8f / 2 - 1).Barcode(str);
        _writer.Position.ShouldBe(new PointF(0, 8f / 2 - 1));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.CenterCenter;
        _writer.MoveTo(6.5f / 2, 8f / 2).Barcode(str);
        _writer.Position.ShouldBe(new PointF(6.5f / 2, 8f / 2));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.RightCenter;
        _writer.MoveTo(6.5f, 8f / 2 + 1).Barcode(str);
        _writer.Position.ShouldBe(new PointF(6.5f, 8f / 2 + 1));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.LeftBottom;
        _writer.MoveTo(0, 8f - 2).Barcode(str);
        _writer.Position.ShouldBe(new PointF(0, 8f - 2));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.CenterBottom;
        _writer.MoveTo(6.5f / 2, 8f - 1).Barcode(str);
        _writer.Position.ShouldBe(new PointF(6.5f / 2, 8f - 1));
        CrossHair(2f);

        _writer.PictureAlignment = PictureAlignment.RightBottom;
        _writer.MoveTo(6.5f, 8f).Barcode(str);
        _writer.Position.ShouldBe(new PointF(6.5f, 8f));
        CrossHair(2f);

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void ColorsSizes()
    {
        _writer.FillColor = Color.Red;
        _writer.QRCode("https://github.com/Shane32/EasyPDF", quietZone: false);
        _writer.QRCodeSize("https://github.com/Shane32/EasyPDF", quietZone: false).ShouldBe(0.8788f, 0.001f);

        _writer.MoveTo(2, 0);
        _writer.LineStyle.JoinStyle = LineJoinStyle.Rounded;
        _writer.LineStyle.Width = 0.0625f * 72;
        _writer.QRCode("https://github.com/Shane32/EasyPDF", quietZone: false);
        _writer.Position.ShouldBe(new PointF(2, 0));

        _writer.MoveTo(0, 2);
        _writer.QRCode("https://github.com/Shane32/EasyPDF", size: 2f, quietZone: false);
        _writer.Position.ShouldBe(new PointF(0, 2));

        _writer.MoveTo(3, 2);
        _writer.QRCode("https://github.com/Shane32/EasyPDF", eccLevel: QRCoder.QRCodeGenerator.ECCLevel.H, size: 2f, quietZone: false);
        _writer.Position.ShouldBe(new PointF(3, 2));

        _writer.MoveTo(0, 4.5f);
        _writer.Barcode("https://github.com");
        _writer.Position.ShouldBe(new PointF(0, 4.5f));

        _writer.MoveTo(0, 5.5f);
        _writer.Barcode("https://github.com", width: 6.5f);
        _writer.Position.ShouldBe(new PointF(0, 5.5f));

        _writer.MoveTo(0, 6.5f);
        _writer.Barcode("https://github.com", height: 1f);
        _writer.Position.ShouldBe(new PointF(0, 6.5f));

        _writer.MoveTo(0, 8);
        _writer.Barcode("https://github.com", width: 6f, height: 1f);
        _writer.Position.ShouldBe(new PointF(0, 8));

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    private void Box(float w, float? h = null)
    {
        using var _ = _writer.SaveState();
        _writer.ForeColor = Color.Blue;
        _writer.LineStyle = new LineStyle(0.02f * 72f, dashStyle: LineDashStyle.Dash);
        _writer.Rectangle(w, h ?? w);
    }
    
    private void CrossHair(float w)
    {
        using var _ = _writer.SaveState();
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.ForeColor = Color.Red;
        _writer.LineStyle = new LineStyle(0.03f * 72f, dashStyle: LineDashStyle.Dash);
        _writer.OffsetTo(-w / 2, 0f).LineTo(w, 0);
        _writer.OffsetTo(-w / 2, -w / 2).LineTo(0, w);
    }

    [Fact]
    public void InvalidBarcodeType()
    {
        Should.Throw<ArgumentOutOfRangeException>(() => _writer.Barcode("test", (BarcodeType)1));
        Should.Throw<ArgumentOutOfRangeException>(() => _writer.BarcodeSize("test", (BarcodeType)1));
    }
}
