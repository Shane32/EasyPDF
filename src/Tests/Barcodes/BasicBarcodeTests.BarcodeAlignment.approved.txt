%PDF-1.4
%\xE2\xE3\xCF\xD3
1 0 obj
<</Length 675/Filter/FlateDecode>>stream
x\x9C\xED\x99\xBD\x8E\xDB0\x0C\x80w=\x85^\xE0\x0CQ\xFFZ\x0B\x14\x05\xBA\xB5\xCDVd\xEA\xB5\x05\x8A\xCBp\xB7\xF4\xF5K\x8Avm\xD1\x074\xBE\xA1\xF0@$q$\x99\x7F\x92\xA8\x0Fa\xFCl\xDE]L\xC8\xB6\xA4l/\x8F\xE6\xFD\xC5|2`\x1D\xBE\x1E\xC0\x16\x8Fog\xBF\xDD\x8C\xB3\x1F\xF1\xF3\xCB\xB8\x09\xECo\xF3\xF5\x8A\xF7\x1F\x8D\xEBr\x9F?\xCC\x8D\x97\x9F\xAF\x8B=\xCF\x06\xF9\xDA\x8Dag\xCA\x16\xDD\xBE|7~\x8Atc\xAAs?bk\xDB\xAF\xA2\x0F0\xF9\xC1\x00\xE4\xF1~\x99\xC6\x01O\xFE\x02*\xCD\xDD\xD1\x9D\xCF\xC2\xBFg\x87\xABB\xF0"\x82\x90z\x04\x9B\x81,%Z\x97XmD1\xC7\xD4c\\\xEF''\xA2H\xEC\x94\x16g\x1E\x90>R\x13\xEB\x90AL<\x87.\xB1\xDA\xC8E\xA8\x14\xE9\xB6\x8C
%\x8F\xE2ML\xBC\xB2\xCBU\xA2\x8E\xE6\xEA\x18r\x1B\xBD\xB7(\xBC79\xC7&\xD7\x11\\\x10!\x80Kb\xDA\xE0\xEA\x98\x1F 7\x10@z\x06\xC8b\xD7\xC1\xCB\xE5\x04\xBF\xF3\xEDw\xBE}\x13+\x02A\xE8\x04\xF6\xBD\x95\xE0\x8C]\x97\x1D\xA2\x93\x11Gy, fi'6\x19\xCD\x9CE\x9BY\xED\xD2\x08R\xEDv62\x99\x8F\xE4&\x9E\xBC\x9B\xE7\x9CJ\xDBc's\x09\x8A\xD8;\x91MP\xDD\xE8\xA5\xEE\xD6\xB7\xF2\xAC\xB7:\xF3q\xBD\xB3\xFF\x03q\xC6\xECAL\xC1\x84\xC7\x10\xB1T\xA7\x1C-]\xAE6N\xB8=L2Ve\xB5/\xAF\xB1m\xCB4Xn\xFB\x09\xF1\x83\xC3\x98c-\xDB~\xBDZ4TQ\xE8\xA1/\xC0\x8D\x0F\xD5S7JC\x04@\xFCz\xFA\xB7\x97\x95\x9C\x9C\x11\xA8\xA5\xF8T|*>\x15\x9F\xFF\x0F\x9Fw\xA1\x0E'm;\xDA\x82\xCB\x0B\xDC|\x88\x1D\x7F\xF4\x8D\x1Bp\x0Cx\xBE\x9FRS\xE2)\xF1\x94xJ\xBC\x93\x11/4\xC6\xD3\xCD\xA4\xE8\xFE\xF2-\xE6\xCA\x1C\xA4\x86G\xEDC\xD0\xC3O\xAB
<\x05\x9E\x02O\x81w6\xE0Q\xE9JD\xEB\xF5\xEC\x8A6f F\x89\x99\xFC\x86\x92\xD6\x17\xFDKP\x81\xA7\xC0S\xE0\x9D\x0ExT\xD3\x12\xD5\xB8\xA8]\xF8F\xD5,c\x90Z!\xBB\xB7\xD4\xB5!\xEA?yJ=\xA5\x9ER\xEFt\xD4\xA3\xBA\x96\xA8\xC6u\xED\xC2\xB7^\xCEv\x16R+\x86\x83\x8F/\xD0\x10\x9AU\xE2)\xF1\x94xJ\xBC\x93\x11\x8F
["Z/lW\xB41\x04\x9DM\xEE\xE0\xA3\x0B\xDEB\xE2\xA4\x02O\x81\xA7\xC0S\xE0\x9D\x0CxT\xD8\x12\xD5\xB8\xB0]\xF8F\xE5,c\x90Z\xA9\x1C|v\xC1\x85-\xFDfT\xEA)\xF5\x94zJ\xBD\x93Q\x8F*P\xA2\x1A\x17\xB6\x0B\xDF\xE8g\x1A\xB3\x90Z9\xDE\xF5\x10\xE3\x0FV\xAB\xE9\xEE
endstream
endobj
3 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]>>/Contents 1 0 R/Parent 2 0 R>>
endobj
2 0 obj
<</Type/Pages/Count 1/Kids[3 0 R]/ITXT(3.3.2.0)>>
endobj
4 0 obj
<</Type/Catalog/Pages 2 0 R>>
endobj
5 0 obj
<<>>
endobj
xref
0 6
0000000000 65535 f 
0000000015 00000 n 
0000000896 00000 n 
0000000757 00000 n 
0000000961 00000 n 
0000001006 00000 n 
trailer
<</Size 6/Root 4 0 R/Info 5 0 